graph/src/main/scala/net/spals/dsa/graph/Graph.scala

```diff
@@ -53,20 +53,63 @@ case class WeightedGraph[T, W](vertices: Set[Vertex[T]], edgeFunc: (Vertex[T], V
   def removeEdge(v1: Vertex[T], v2: Vertex[T]): WeightedGraph[T, W] = {
     val removeEdgeFunc: (Vertex[T], Vertex[T]) => Option[W] =
       (vx, vy) => (vx, vy) match {
-        case (vx, vy) if (vx == v1 && vy == v2) || (vx == v2 && vy == v1) => None
-        case _ => edgeFunc.apply(vx, vy)
-      }
+          case (`vx`, `vy`) if (vx == v1 && vy == v2) || (vx == v2 && vy == v1) => None
+          case _ => edgeFunc.apply(vx, vy)
+        }
 
     WeightedGraph(vertices, removeEdgeFunc)
   }
 
   def removeVertex(v: Vertex[T]): WeightedGraph[T, W] = {
     val removeVertexFunc: (Vertex[T], Vertex[T]) => Option[W] =
       (vx, vy) => (vx, vy) match {
-        case (vx, vy) if vx == v || vy == v => None
+        case (`vx`, `vy`) if vx == v || vy == v => None
         case _ => edgeFunc.apply(vx, vy)
       }
 
     WeightedGraph(vertices - v, removeVertexFunc)
   }
 }
+
+object SimpleGraph {
+
+  def empty[T]: SimpleGraph[T] = SimpleGraph(WeightedGraph.empty[T, Int])
+}
+
+/**
+  * Defined as a weighted graph with all edge weights as a unit, i.e. the integer value of 1.
+  */
+case class SimpleGraph[T](graph: WeightedGraph[T, Int]) {
```

Comment:
In scala, interfaces are called `trait`s. But you can include code in them. So they are a little like a traditional Java abstract class (Java 8 allows code in interfaces with the `default` keyword).